java-support/src/main/scala/io/cloudstate/javasupport/impl/eventsourced/EventSourcedImpl.scala

```diff
@@ -99,18 +123,17 @@ final class EventSourcedImpl(_system: ActorSystem,
         case (Seq(EventSourcedStreamIn(InInit(init), _)), source) =>
           source.via(runEntity(init))
         case _ =>
-          // todo better error
-          throw new RuntimeException("Expected Init message")
+          throw ProtocolException("Expected Init message")
       }
       .recover {
-        case e =>
-          // FIXME translate to failure message
-          throw e
+        case EntityException(id, message) => failure(id, message)
+        case ProtocolException(id, message) => failure(id, "Protocol error: " + message)
```

Comment:
Both cases produce a `EventSourcedStreamOut.OutFailure.Failure` is that correct?

Don't we send:

`EntityException` => `EventSourcedStreamOut.Message.Failure`
`ProtocolException`  => `ClientAction.Action.Failure`

Reply:
^^^ edited! got it wrong at first :)

Reply:
Any exceptions at this point (in the `recover` for the akka stream) are going to be converted into entity failures — these are all unexpected, while `context.fail` in the java-support throws a `FailInvoked` which is caught when invoking handlers and becomes a failure in the reply. Other exceptions thrown anywhere in the akka stream will fail the stage they occur in and then cancel (upstream) and complete (downstream) with the failure. The `recover` allows the exception to be converted into a final message before the stream is closed.

So the behaviour now is that any unexpected exception (not `FailInvoked`) in the service creates an entity failure message plus closes the stream, and on the proxy side the failure is reported to the client with a generic message and the entity is stopped and stream closed.

I distinguished the protocol exception — some failure in the protocol interaction with the proxy — from exceptions in the entity. The only differences in the end are that the entity exception message is left as is for the failure description, the protocol exception has "Protocol error:" and any other exceptions have "Unexpected failure:" — but maybe it's more understandable in the impl to just have one exception type for entity failures. The special exception type allows the command id to be added when available, which is its main purpose.